pallets/ds-accounts/src/lib.rs

```diff
@@ -232,18 +238,24 @@ decl_module! {
             let who = ensure_signed(origin)?;
             ensure!(Self::account_is_registrar(&who), Error::<T>::NotAuthorized);
 
-            // Update storage.
-            AccountRegistry::<T>::mutate(&account, |acc|{
+            let update_storage_result = AccountRegistry::<T>::mutate(&account, |acc| -> dispatch::DispatchResult {
+                ensure!(!AccountOf::<T>::is_pilot(acc), Error::<T>::AlreadyRegistered);
                 debug::info!("register_pilot: roles:{:?} create_time={:?}", acc.roles, acc.create_time);
+
                 acc.roles = acc.roles | PILOT_ROLE.into();
                 if acc.create_time.is_zero() {
                     acc.create_time = <pallet_timestamp::Module<T>>::get();
                 }
                 acc.managed_by = who.clone();
+
+                Ok(())
             });
 
-            Self::deposit_event(RawEvent::PilotRegistered(who, account));
-            Ok(())
+            if update_storage_result.is_ok() {
```

Comment:
It makes sence. You can leave it like this. As well we have only one mutate method,  Self::deposit_event can be put just before  Ok(()) of the closure , for simplicity.